### PR TITLE
Re-design functionalization to minimize miss in-placing when args are views.

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134315
* __->__ #133045
* #134248

Previous functionalization fails to re-inplace arguments when they are view over other tensors. 
see issue https://github.com/pytorch/pytorch/issues/131192

The new functionalization is easier to re-inplace when views are used it works as the following:

**A) Functionalizations pass**

consider a program 
```
t = [..]
y = view(x)
z = view (x)
foo(y, z) // where y and z are mutable inputs to custom op foo
return (t, y, z)
```

1) When we perform functionalization we track for each arg its base (if its a view, otherwise we consider itself to be
itself base).  we add those to a list _all_bases=[t] in this case and pass it to ``auto_functionalize`'

2) We also add an arg for each mutated arg that maps the arg to it base, _y_base= t, and _z_base =t in the example above. 
 
3) The new output of ``auto_functionalize``, is the args in _all_bases. 

4) If the mutated input is an array of tensors, then _x_base=[b1, b2, b3] is an array of bases.
 
5) We inform the cpp functionlization that t = t' has changed, and ask it whenever it encounters a a view over t later in the code then it  needs to regenerate the view from the new  t' value. This is done by calling replace(t, t'), commit_update(t) and sync(t) sequence.

For example for the function above the program after the functionalization is:
 
```
t = [..]
y = t[0]
z = t[1]
t' = auto_functionalize(foo, y, z, _y_base=t , z_base=t, _all_bases=t)
y' = t'[0]
z' = t'[1]
return (t', y', z')
```

**B) Semantics of  auto_functionalize**
The new semantics of auto_functionalize is as the following:
```
1. copy all mutated inputs into new variables input'
2. call the custom op on the copies of the mutated inputs. 
3. for each base in _all_base 
      base' = base
      for each mutated input' 
          if _input_base != base
             continue
              
          if base' is  input :
              base' = input'
           elseif  input is view on base' 
              base' = alias'.as_strided_scatter(input',input'.size(), input'.stride(), input'.storage_offset()) 
              
4. return [base'...] ## all the new base' values
```
 
**C) Re-inplace pass**
the following changes are applied to re-inplace pass:
1. if the argument to a function is a view of another input, we do not reinplace
 if we do not find the copy node that reflects the mutation to the base. 
2. for each argument we can inplace if  non of the argument or any of its aliases, is used 
after the auto_functionalize (before the copy node if exists ). 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

File: torch/_higher_order_ops/auto_functionalize.py [hunk 1]
```diff
@@ -242,13 +244,33 @@ def do_auto_functionalize(
...
+    # for each arg, identify the items that
+    with ctx.redispatch_to_next():
+        # The output have the the following structure:
+        # - op output
+        # - mutated 0, alias 0, alias 1, alias 2, alias 3...etc
+        # - mutated kwargs
+        # We only need the first one, so we slice it.
+        print("inputs")
+        print(unwrapped_kwargs)
+        unwrapped_outs = auto_functionalized(
```
Comment (laithsakka) [inline torch/_higher_order_ops/auto_functionalize.py:271]: passing arg_to_aliased and all_aliased result in the following error:
passing them empty is fine
```

Traceback (most recent call last):
  File "/home/lsakka/pytorch/example.py", line 31, in <module>
    f(x)
  File "/home/lsakka/pytorch/torch/_dynamo/eval_frame.py", line 464, in _fn
    return fn(*args, **kwargs)
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 1224, in __call__
    return self._torchdynamo_orig_callable(
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 514, in __call__
    return _compile(
  File "/home/lsakka/.conda/envs/myenv/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 896, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 662, in compile_inner
    return _compile_inner(code, one_graph, hooks, transform)
  File "/home/lsakka/pytorch/torch/_utils_internal.py", line 85, in wrapper_function
    return StrobelightCompileTimeProfiler.profile_compile_time(
  File "/home/lsakka/pytorch/torch/_strobelight/compile_time_profiler.py", line 129, in profile_compile_time
    return func(*args, **kwargs)
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 697, in _compile_inner
    out_code = transform_code_object(code, transform)
  File "/home/lsakka/pytorch/torch/_dynamo/bytecode_transformation.py", line 1322, in transform_code_object
    transformations(instructions, code_options)
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 209, in _fn
    return fn(*args, **kwargs)
  File "/home/lsakka/pytorch/torch/_dynamo/convert_frame.py", line 631, in transform
    tracer.run()
  File "/home/lsakka/pytorch/torch/_dynamo/symbolic_convert.py", line 2722, in run
    super().run()
  File "/home/lsakka/pytorch/torch/_dynamo/symbolic_convert.py", line 957, in run
    while self.step():
  File "/home/lsakka/pytorch/torch/_dynamo/symbolic_convert.py", line 869, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/home/lsakka/pytorch/torch/_dynamo/symbolic_convert.py", line 2913, in RETURN_VALUE
    self._return(inst)
  File "/home/lsakka/pytorch/torch/_dynamo/symbolic_convert.py", line 2898, in _return
    self.output.compile_subgraph(
  File "/home/lsakka/pytorch/torch/_dynamo/output_graph.py", line 1133, in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
  File "/home/lsakka/.conda/envs/myenv/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/lsakka/pytorch/torch/_dynamo/output_graph.py", line 1360, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/home/lsakka/pytorch/torch/_dynamo/output_graph.py", line 1407, in call_user_compiler
    return self._call_user_compiler(gm)
  File "/home/lsakka/pytorch/torch/_dynamo/output_graph.py", line 1437, in _call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/home/lsakka/pytorch/torch/_dynamo/repro/after_dynamo.py", line 129, in __call__
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/lsakka/pytorch/torch/_dynamo/repro/after_dynamo.py", line 129, in __call__
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/lsakka/pytorch/torch/__init__.py", line 2231, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "/home/lsakka/.conda/envs/myenv/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 1488, in compile_fx
    return aot_autograd(
  File "/home/lsakka/pytorch/torch/_dynamo/backends/common.py", line 69, in __call__
    cg = aot_module_simplified(gm, example_inputs, **self.kwargs)
  File "/home/lsakka/pytorch/torch/_functorch/aot_autograd.py", line 1019, in aot_module_simplified
    compiled_fn = dispatch_and_compile()
  File "/home/lsakka/pytorch/torch/_functorch/aot_autograd.py", line 1008, in dispatch_and_compile
    compiled_fn, _ = create_aot_dispatcher_function(
  File "/home/lsakka/pytorch/torch/_functorch/aot_autograd.py", line 429, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(flat_fn, flat_args, aot_config)
  File "/home/lsakka/pytorch/torch/_functorch/aot_autograd.py", line 730, in _create_aot_dispatcher_function
    compiled_fn, fw_metadata = compiler_fn(
  File "/home/lsakka/pytorch/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 178, in aot_dispatch_base
    compiled_fw = compiler(fw_module, updated_flat_args)
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 1327, in fw_compiler_base
    return _fw_compiler_base(model, example_inputs, is_inference)
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 1394, in _fw_compiler_base
    return inner_compile(
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 457, in compile_fx_inner
    return _compile_fx_inner(*args, **kwargs)
  File "/home/lsakka/pytorch/torch/_dynamo/repro/after_aot.py", line 85, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "/home/lsakka/pytorch/torch/_inductor/debug.py", line 313, in inner
    return fn(*args, **kwargs)
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 446, in wrapper
    return fn(*args, **kwargs)
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 644, in _compile_fx_inner
    compiled_graph = FxGraphCache.load(
  File "/home/lsakka/pytorch/torch/_inductor/codecache.py", line 1314, in load
    compiled_graph = compile_fx_fn(
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 554, in codegen_and_compile
    compiled_graph = fx_codegen_and_compile(gm, example_inputs, **fx_kwargs)
  File "/home/lsakka/.conda/envs/myenv/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/lsakka/pytorch/torch/_inductor/compile_fx.py", line 836, in fx_codegen_and_compile
    graph.run(*example_inputs)
  File "/home/lsakka/pytorch/torch/_inductor/graph.py", line 768, in run
    return super().run(*args)
  File "/home/lsakka/pytorch/torch/fx/interpreter.py", line 147, in run
    self.env[node] = self.run_node(node)
  File "/home/lsakka/pytorch/torch/_inductor/graph.py", line 1276, in run_node
    result = super().run_node(n)
  File "/home/lsakka/pytorch/torch/fx/interpreter.py", line 204, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/home/lsakka/pytorch/torch/_inductor/graph.py", line 1074, in get_attr
    return tensor(value.tolist(), dtype=value.dtype, device=value.device)
  File "/home/lsakka/pytorch/torch/_subclasses/fake_tensor.py", line 881, in tolist
    torch._check_is_size(s)
  File "/home/lsakka/pytorch/torch/__init__.py", line 1571, in _check_is_size
    _check(i >= 0, message)
  File "/home/lsakka/pytorch/torch/__init__.py", line 1557, in _check
    _check_with(RuntimeError, cond, message)
  File "/home/lsakka/pytorch/torch/__init__.py", line 1520, in _check_with
    if expect_true(cond):
  File "/home/lsakka/pytorch/torch/fx/experimental/symbolic_shapes.py", line 986, in expect_true
    return a.node.expect_true(frame.f_code.co_filename, frame.f_lineno)
  File "/home/lsakka/pytorch/torch/fx/experimental/sym_node.py", line 476, in expect_true
    return self.shape_env.defer_runtime_assert(
  File "/home/lsakka/pytorch/torch/fx/experimental/recording.py", line 262, in wrapper
    return retlog(fn(*args, **kwargs))
  File "/home/lsakka/pytorch/torch/fx/experimental/symbolic_shapes.py", line 5451, in defer_runtime_assert
    assert not self.runtime_asserts_frozen, expr
AssertionError: u0 >= 0

```

**Comment by laithsakka** [comment]
changing auto_functionalized_proxy to avoid hitting the issue for now but thats just by trail and error. 
I probably need they motivation behind all those auto_functionalized_X functions better. 
```
@auto_functionalized.py_impl(ProxyTorchDispatchMode)
def auto_functionalized_proxy(
    mode,
    _mutable_op: torch._ops.OpOverload,
    _arg_to_aliased= None,
    _all_aliased =None,
    **kwargs: Dict[str, Any],
) -> Tuple[Any, Tuple[Tensor, ...]]:
    if not mode.enable_tracing:
        return auto_functionalized(_mutable_op,_arg_to_aliased, _all_aliased, **kwargs)

    with disable_proxy_modes_tracing():
        out = auto_functionalized(_mutable_op, _arg_to_aliased, _all_aliased, **kwargs)

    proxy_kwargs = pytree.tree_map(mode.tracer.unwrap_proxy, kwargs)
    out_proxy = mode.tracer.create_proxy(
        "call_function",
        auto_functionalized,
        (_mutable_op,None, None,),
        proxy_kwargs,
    )
    result = track_tensor_tree(out, out_proxy, constant=None, tracer=mode.tracer)
    return result
 ```

**Comment by laithsakka** [comment]
ah if i put it back and just use aot_autograd 
```
NotImplementedError: There was no rule registered for HOP auto_functionalized and subclass <class 'torch._subclasses.fake_tensor.FakeTensor'>. We recommend filing an issue.
While executing %auto_functionalized : [num_users=2] = call_function[target=torch.ops.higher_order.auto_functionalized](args = (mylib.foo.default, {x: [%_tensor_constant0, %_tensor_constant1, %_tensor_constant2], y: [%_tensor_constant0_1, %_tensor_constant1_1, %_tensor_constant2_1]}, [%_tensor_constant0_2, %_tensor_constant2_2, %_tensor_constant1_2]), kwargs = {x: %select, y: %select_1})
Original traceback:
  File "/home/lsakka/pytorch/example.py", line 28, in f
    foo(a, b)
  File "/home/lsakka/pytorch/torch/_library/custom_ops.py", line 664, in __call__
    return self._opoverload(*args, **kwargs)
```
note the _tensor_constant0.. etc
those unfortuently are not referring to select, select_1 which is what we want .

**Comment by laithsakka** [comment]
> ah if i change auto_functionalized_proxy  back and use aot_autograd as backend. I get this error 
> 
> ```
> NotImplementedError: There was no rule registered for HOP auto_functionalized and subclass <class 'torch._subclasses.fake_tensor.FakeTensor'>. We recommend filing an issue.
> While executing %auto_functionalized : [num_users=2] = call_function[target=torch.ops.higher_order.auto_functionalized](args = (mylib.foo.default, {x: [%_tensor_constant0, %_tensor_constant1, %_tensor_constant2], y: [%_tensor_constant0_1, %_tensor_constant1_1, %_tensor_constant2_1]}, [%_tensor_constant0_2, %_tensor_constant2_2, %_tensor_constant1_2]), kwargs = {x: %select, y: %select_1})
> Original traceback:
>   File "/home/lsakka/pytorch/example.py", line 28, in f
>     foo(a, b)
>   File "/home/lsakka/pytorch/torch/_library/custom_ops.py", line 664, in __call__
>     return self._opoverload(*args, **kwargs)
> ```
> 
> note the _tensor_constant0.. etc those unfortuently are not referring to select, select_1 which is what we want .

**Comment by laithsakka** [comment]
also this graph is not what we want for those tensors.
   ```
    %arg0_1 : [num_users=4] = placeholder[target=arg0_1]
    %select : [num_users=1] = call_function[target=torch.ops.aten.select.int](args = (%arg0_1, 0, 0), kwargs = {})
    %select_1 : [num_users=1] = call_function[target=torch.ops.aten.select.int](args = (%arg0_1, 0, 1), kwargs = {})
    %_tensor_constant0 : [num_users=1] = get_attr[target=_tensor_constant0]
    %_tensor_constant1 : [num_users=1] = get_attr[target=_tensor_constant1]
    %_tensor_constant2 : [num_users=1] = get_attr[target=_tensor_constant2]
    %_tensor_constant0_1 : [num_users=1] = get_attr[target=_tensor_constant0]
    %_tensor_constant1_1 : [num_users=1] = get_attr[target=_tensor_constant1]
    %_tensor_constant2_1 : [num_users=1] = get_attr[target=_tensor_constant2]
    %_tensor_constant0_2 : [num_users=1] = get_attr[target=_tensor_constant0]
    %_tensor_constant1_2 : [num_users=1] = get_attr[target=_tensor_constant1]
    %_tensor_constant2_2 : [num_users=1] = get_attr[target=_tensor_constant2]
    %auto_functionalized : [num_users=2] = call_function[target=torch.ops.higher_order.auto_functionalized](args = (mylib.foo.default, {x: [%_tensor_constant0, %_tensor_constant1, %_tensor_constant2], y: [%_tensor_constant0_1, %_tensor_constant1_1, %_tensor_constant2_1]}, [%_tensor_constant0_2, %_tensor_constant1_2, %_tensor_constant2_2]), kwargs = {x: %select, y: %select_1})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%auto_functionalized, 1), kwargs = {})
    %getitem_2 : [num_users=1] = call_function[target=operator.getitem](args = (%auto_functionalized, 2), kwargs = {})
    %select_scatter : [num_users=1] = call_function[target=torch.ops.aten.select_scatter.default](args = (%arg0_1, %getitem_1, 0, 0), kwargs = {})
    %select_scatter_1 : [num_users=1] = call_function[target=torch.ops.aten.select_scatter.default](args = (%select_scatter, %getitem_2, 0, 1), kwargs = {})
    %copy_ : [num_users=0] = call_function[target=torch.ops.aten.copy_.default](args = (%arg0_1, %select_scatter_1), kwargs = {})
    ```

File: torch/_higher_order_ops/auto_functionalize.py [hunk 2]
```diff
@@ -294,8 +321,10 @@ def sync_update(o, orig_arg):
 
 
 @auto_functionalized.py_functionalize_impl
-def auto_functionalized_func(ctx, _mutable_op, **kwargs):
+def auto_functionalized_func(ctx, _mutable_op, _arg_to_aliased, _all_aliased, **kwargs):
     unwrapped_kwargs = ctx.unwrap_tensors(kwargs)
     with ctx.redispatch_to_next():
-        result = auto_functionalized(_mutable_op, **unwrapped_kwargs)
+        result = auto_functionalized(
+            _mutable_op, _arg_to_aliased, _all_aliased, **unwrapped_kwargs
```
Comment (laithsakka) [inline torch/_higher_order_ops/auto_functionalize.py:328]: ah this is wrong. need to append those to unwrapped_kwargs, but idk if that offend users!

File: test/dynamo/test_misc.py [hunk 3]
```diff
@@ -931,6 +932,265 @@ def forward(self, arg0_1: "f32[3][1]cpu", arg1_1: "f32[3][1]cpu", arg2_1: "f32[3
...
+        as_strided_default_1: "f32[][]cpu" = torch.ops.aten.as_strided.default(clone_default, [], [], 1);  clone_default = None
+        foo_default = torch.ops.mylib.foo.default(select, as_strided_default_1);  select = foo_default = None
+        as_strided_scatter_default: "f32[2][1]cpu" = torch.ops.aten.as_strided_scatter.default(arg0_1, as_strided_default_1, [], [], 1);  as_strided_default_1 = None
+
+        copy_: "f32[2][1]cpu" = torch.ops.aten.copy_.default(arg0_1, as_strided_scatter_default);  arg0_1 = copy_ = None
+
+        select_2: "f32[][]cpu" = torch.ops.aten.select.int(as_strided_scatter_default, 0, 0)
+        select_3: "f32[][]cpu" = torch.ops.aten.select.int(as_strided_scatter_default, 0, 1);  as_strided_scatter_default = None
+        return (select_2, select_3)""",
+                    ignore_comments=True,
```
Comment (laithsakka) [inline test/dynamo/test_misc.py:1188]: I will add two more tests one with None, and one with mutable list.

File: torch/_higher_order_ops/auto_functionalize.py [hunk 4]
```diff
@@ -234,23 +315,66 @@ def do_auto_functionalize(
...
+    # List of the name of args that get mutated (according to the schema)
+    mutable_args_names = get_mutable_arg_names(op)
+
+    all_basis = set()
+    all_basis_addresses: set[int] = set()
+    basis: Dict[str, Any] = {}
+
+    for arg_name in mutable_args_names:
+        arg = normalized_kwargs[arg_name]
+        if arg is None:
```
Comment (laithsakka) [inline torch/_higher_order_ops/auto_functionalize.py:333]: check if shall do 
basis[arg_name] = None

File: torch/testing/_internal/common_utils.py [hunk 5]
```diff
@@ -2860,7 +2860,35 @@ def assertLeaksNoCudaTensors(self, name=None):
     def enforceNonDefaultStream(self):
         return CudaNonDefaultStream()
 
-    def assertExpectedInline(self, actual, expect, skip=0):
+    def _remove_ansi_escape(self, input):
+        # 7-bit C1 ANSI sequences
```
Comment (laithsakka) [inline torch/testing/_internal/common_utils.py:2864]: split this into its own PR.

**Comment by laithsakka** [comment]
there is one failure which i fix in 
https://github.com/pytorch/pytorch/pull/134315
if we agree we want to fix it.

File: test/export/test_export.py [hunk 6]
```diff
@@ -6021,7 +6021,7 @@ def forward(self, x):
             """\
 def forward(self, x):
     cos = torch.ops.aten.cos.default(x)
-    auto_functionalized = torch.ops.higher_order.auto_functionalized(torch.ops.testlib.foo.default, x = x, z = cos);  x = cos = None
+    auto_functionalized = torch.ops.higher_order.auto_functionalized(torch.ops.testlib.foo.default, x = x, z = cos, _x_base = x, _z_base = cos, _all_bases = [x, cos]);  x = cos = None
```
Comment (zou3519) [inline test/export/test_export.py:6024]: Why do we need `_x_base`, `_z_base`, and `_all_bases` ?

File: torch/_higher_order_ops/auto_functionalize.py [hunk 7]
```diff
@@ -58,8 +61,7 @@ def __init__(self) -> None:
         super().__init__("auto_functionalized")
 
     def __call__(
-        self,
-        /,
+        self_,  # noqa: B902
```
Comment (zou3519) [inline torch/_higher_order_ops/auto_functionalize.py:64]: Why the change?

File: test/export/test_export.py (hunk 6 above)
Comment (laithsakka) [reply test/export/test_export.py:6024]: I am changing it to 
[x', y'] = auto_functionalize( foo, x, y, _all_bases=[x,y], _observe_mutation_from=[[arg0],[arg1])
see https://github.com/pytorch/pytorch/pull/134315

File: torch/_higher_order_ops/auto_functionalize.py (hunk 7 above)
Comment (laithsakka) [reply torch/_higher_order_ops/auto_functionalize.py:64]: local linter i guess will reverse it .

File: test/export/test_export.py (hunk 6 above)
Comment (laithsakka) [reply test/export/test_export.py:6024]: Ah back to your question, _all_bases removes the repetitions and specify the order of the outputs.

File: torch/_higher_order_ops/auto_functionalize.py [hunk 8]
```diff
@@ -126,38 +128,119 @@ def auto_functionalized_dense(
     _only_clone_these_tensors: Optional[Tuple[str, ...]] = None,
     **kwargs: Any,
 ) -> Tuple[Any, Tuple[Tensor, ...]]:
+    _all_bases: List[Tensor] = kwargs.pop("_all_bases", [])
```
Comment (zou3519) [inline torch/_higher_order_ops/auto_functionalize.py:131]: I just realized we need to handle the following case in this PR: if the base has no view relationship with the tensor (maybe due to a graph pass), then we don't know what the right thing to do is.

So, in addition to recording x_base for each x, we also need to record x.stride() and x.storage_offset(). This needs to happen in this PR (sorry I didn't realize earlier)

Comment (laithsakka) [reply torch/_higher_order_ops/auto_functionalize.py:131]: no worries lets chat about this and about the change that i made to this. to handle test_multi_output_intermediate when you have time. I would like to understand better this comment.

Comment (zou3519) [reply torch/_higher_order_ops/auto_functionalize.py:131]: Under this design we wouldn't need to pass `x` anymore (it doesn't matter what's in x!). Instead we can pass (x_sizes, x_strides, x_storage_offset) instead of x.

Also, the semantics of auto_functionalized become:

```py
x_base = maybe_clone(x_base)  # if we can't reinplace it
x = x_base.as_strided(x_sizes, x_strides, x_storage_offset)
mutable_op(x)
```
But we also want to somehow not emit an as_strided when it's unnecessary

**Review by zou3519 — COMMENTED** [review]
I'll be back to review the rest, but we might need to change how the auto_functionalized node works (+ had some comments to simplify it).

File: torch/_higher_order_ops/auto_functionalize.py [hunk 9]
```diff
@@ -234,23 +318,66 @@ def do_auto_functionalize(
...
+            continue
+
+        if isinstance(arg, list):
+            basis[arg_name] = []
+            for tensor in arg:
+                if tensor is None:
+                    basis[arg_name].append(None)
+                    continue
+
+                base = tensor if tensor._base is None else tensor._base
```
Comment (zou3519) [inline torch/_higher_order_ops/auto_functionalize.py:346]: We should check if tensor._base is available under torch.inference_mode(). If it's not then... we're gonna need to add more infra to keep track of bases.

Comment (laithsakka) [reply torch/_higher_order_ops/auto_functionalize.py:346]: with torch.inference_mode():
    x = torch.randn(3)
    f(x)
